sbi/utils/score_utils.py

```diff
@@ -0,0 +1,597 @@
+    """Marginalize a MixtureSameFamily distribution.
+
+    Args:
+        p: The prior MixtureSameFamily distribution.
+        m: The scaling factor m.
+        s: The standard deviation of the noise s.
+
+    Returns:
+        The marginal MixtureSameFamily distribution.
+    """
+    return MixtureSameFamily(
+        p.mixture_distribution, marginalize(p.component_distribution, m, s)
+    )
+
+
+def marginalize_gaussian(p: Normal, m: Tensor, s: Tensor) -> Normal:
+    """Marginalize a Gaussian distribution.
+
+    Args:
+        p: The prior Gaussian distribution.
+        m: The scaling factor.
+        s: The standard deviation of the noise.
+
+    Returns:
+        The marginal Gaussian distribution.
+    """
+    mean_0 = p.loc
+    std_0 = p.scale
+
+    # Calculate the marginal mean and variance
+    marginal_mean = m * mean_0
+    marginal_variance = (m * std_0) ** 2 + s**2
+    marginal_std = torch.sqrt(marginal_variance)
+
+    return Normal(marginal_mean, marginal_std)
+
+
+def marginalize_multivariate_gaussian(
+    p: MultivariateNormal, m: Tensor, s: Tensor
+) -> MultivariateNormal:
+    """Marginalize a multivariate Gaussian distribution.
+
+    Given an observation model xₜ = m * x + ε with independent noise ε ~ N(0, s² I),
+    where x ~ p, return the marginal distribution p(xₜ) as a MultivariateNormal.
+
+    Args:
+        p: The prior multivariate Gaussian distribution.
+        m: The scaling factor.
+        s: The standard deviation of the noise.
+
+    Returns:
+        The marginal multivariate Gaussian distribution p(xₜ).
+    """
+    mean_0 = p.loc
+    cov_0 = p.covariance_matrix
+
+    marginal_mean = m * mean_0
+    marginal_cov = (m**2) * cov_0 + s**2 * torch.eye(
+        mean_0.shape[-1],
+        dtype=cov_0.dtype,  # type: ignore
+        device=cov_0.device,  # type: ignore
+    )
+
+    return MultivariateNormal(marginal_mean, covariance_matrix=marginal_cov)
+
+
+def marginalize_uniform(
+    p: Uniform | BoxUniform, m: Tensor, s: Tensor
+) -> 'UniformNormalConvolution':
+    """Marginalize a uniform distribution.
+
+    Args:
+        p: The prior uniform distribution.
+        m: The scaling factor.
+        s: The standard deviation of the noise.
+
+    Returns:
+        The marginal uniform distribution.
+    """
+
+    return UniformNormalConvolution(p.low, p.high, m, s)  # type: ignore
+
+
+def marginalize_empirical(
+    p: Distribution, m: Tensor, s: Tensor, num_particles: int = 5000
+) -> MixtureSameFamily:
+    """Marginalize an empirical distribution.
+
+    Args:
+        p: The prior empirical distribution.
+        m: The scaling factor m.
+        s: The standard deviation of the noise s.
+        num_particles: Number of particles to sample. Defaults to 5000.
+
+    Returns:
+        The marginal empirical distribution.
+    """
+    gmm_approx = fit_gmm(p)
+    return marginalize_mixture(gmm_approx, m, s)
+
+
+# Special distributions:
+class UniformNormalPosterior(Distribution):
```

Comment:
does it make sense to move this to a more general utils? e.g., during linear gaussian testing we need a uniform prior Gaussian likelihood gt posterior.

Reply:
please add types throughout - thanks!

Reply:
I guess. But this is a very constrained "class" i.e. only supports sclar scaling and std per dimension ...